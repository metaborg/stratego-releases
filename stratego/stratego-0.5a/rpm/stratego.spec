# 	$Id: stratego.spec,v 1.1 2000/06/20 00:12:00 visser Exp $

%define name        stratego
%define version     0.4.15
%define release     ev

Summary: Stratego
Name: stratego
Version: %{version}
Release: %{release}
Copyright: ev
Group: Programming/Statego
Source0: stratego-%{version}.tar.gz		
BuildRequires: aterm >= 0.24		

%description
Summary: Stratego

%prep
%setup -q 
./configure --prefix=/usr/local --with-aterm=/usr/local		

%build
make

%install
make install

%clean
rm -rf $RPM_BUILD_ROOT

%files
%defattr(-,root,root)
/usr/local/lib/libstratego.a
/usr/local/lib/libstratego.dbg.a
/usr/local/lib/libstratego.opt.a
/usr/local/lib/libstratego.prof.a
/usr/local/lib/libstratego-lib.a
/usr/local/lib/libstratego-lib.dbg.a
/usr/local/lib/libstratego-lib.opt.a
/usr/local/lib/libstratego-lib.prof.a
/usr/local/include/stratego.h
/usr/local/include/svm.h
/usr/local/include/debug.h
/usr/local/include/aterm-extension.h
/usr/local/include/options.h
/usr/local/include/util.h
/usr/local/include/SL-tuples.h
/usr/local/include/SL.h
/usr/local/include/stratego-ext.h
/usr/local/include/stratego-lib.h
/usr/local/include/tables.h
/usr/local/share/stratego/Annotations.r
/usr/local/share/stratego/free-variables.r
/usr/local/share/stratego/list-set-test.r
/usr/local/share/stratego/sglr.r
/usr/local/share/stratego/BinSet.r
/usr/local/share/stratego/gpp.r
/usr/local/share/stratego/list-set.r
/usr/local/share/stratego/share-misc.r
/usr/local/share/stratego/Share.r
/usr/local/share/stratego/int-list.r
/usr/local/share/stratego/list-sort.r
/usr/local/share/stratego/share-test.r
/usr/local/share/stratego/abox-ext.r
/usr/local/share/stratego/integers.r
/usr/local/share/stratego/list-zip.r
/usr/local/share/stratego/share.r
/usr/local/share/stratego/abox.r
/usr/local/share/stratego/io-text.r
/usr/local/share/stratego/list.r
/usr/local/share/stratego/simple-traversal.r
/usr/local/share/stratego/annotations.r
/usr/local/share/stratego/io.r
/usr/local/share/stratego/memo.r
/usr/local/share/stratego/string.r
/usr/local/share/stratego/basic.r
/usr/local/share/stratego/iteration.r
/usr/local/share/stratego/option-test.r
/usr/local/share/stratego/substitution.r
/usr/local/share/stratego/bin-tree-set.r
/usr/local/share/stratego/lib.r
/usr/local/share/stratego/option.r
/usr/local/share/stratego/tables.r
/usr/local/share/stratego/bin-tree.r
/usr/local/share/stratego/list-basic.r
/usr/local/share/stratego/options.r
/usr/local/share/stratego/term.r
/usr/local/share/stratego/conditional.r
/usr/local/share/stratego/list-cons.r
/usr/local/share/stratego/pack-graph.r
/usr/local/share/stratego/test.r
/usr/local/share/stratego/env-traversal.r
/usr/local/share/stratego/list-filter.r
/usr/local/share/stratego/pack-modules.r
/usr/local/share/stratego/thread.r
/usr/local/share/stratego/exec.r
/usr/local/share/stratego/list-index.r
/usr/local/share/stratego/pack.r
/usr/local/share/stratego/time.r
/usr/local/share/stratego/file.r
/usr/local/share/stratego/list-lookup.r
/usr/local/share/stratego/parse-options.r
/usr/local/share/stratego/tuple.r
/usr/local/share/stratego/fixpoint-traversal.r
/usr/local/share/stratego/list-misc.r
/usr/local/share/stratego/rename.r
/usr/local/share/stratego/unification.r
/usr/local/share/stratego/Stratego.Make
/usr/local/bin/backend
/usr/local/bin/frontend
/usr/local/bin/extract
/usr/local/bin/inline
/usr/local/bin/matching-tree
/usr/local/bin/optimizer
/usr/local/bin/postprocess
/usr/local/bin/pp-instructions
/usr/local/bin/proto-sc
/usr/local/bin/parse-stratego
/usr/local/bin/parse-mod
/usr/local/bin/pack-stratego

%changelog
* Sat Jun 17 2000 marco <marco@wildape.com>
- First release
