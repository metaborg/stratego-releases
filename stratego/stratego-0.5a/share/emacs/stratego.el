(setq auto-mode-alist (cons '("\\.r\\'" . stratego-mode) auto-mode-alist))
(setq auto-mode-alist (cons '("\\.ss\\'" . stratego-mode) auto-mode-alist))

(defvar stratego-mode-hook nil
  "Normal hook run when entering Stratego mode.")

(defvar stratego-mode-map nil
  "Keymap for Stratego mode.")

;;;(if stratego-mode-map
;;;    nil
;;;    (progn 
(setq stratego-mode-map (make-sparse-keymap))
;;;	   (define-key stratego-mode-map "\C-m"
;;;	     'reindent-then-newline-and-indent)
;;;	   (define-key stratego-mode-map "\C-j" 'newline)
	   (define-key stratego-mode-map [tab] 'tab-to-tab-stop)
;;;))

(defvar stratego-mode-syntax-table nil
  "Syntax table for Stratego mode.")

(if stratego-mode-syntax-table
    nil
    (progn
      (setq stratego-mode-syntax-table (make-syntax-table))
      (modify-syntax-entry ?( "() 1" stratego-mode-syntax-table)
      (modify-syntax-entry ?) ")( 4" stratego-mode-syntax-table)
;;;      (modify-syntax-entry ?< "(>" stratego-mode-syntax-table)
;;;      (modify-syntax-entry ?> ")<" stratego-mode-syntax-table)
      (modify-syntax-entry ?* ". 23" stratego-mode-syntax-table)
      (modify-syntax-entry ?\' "w" stratego-mode-syntax-table)
      (modify-syntax-entry ?\\ "\\" stratego-mode-syntax-table)))

(defvar sdf2-mode-syntax-table t
  "Syntax table for SDF2 mode.")

(if sdf2-mode-syntax-table
    (progn
  (setq sdf2-mode-syntax-table (make-syntax-table))
  (modify-syntax-entry ?% "<" sdf2-mode-syntax-table)
  (modify-syntax-entry ?\n ">" sdf2-mode-syntax-table)
  (modify-syntax-entry ? ">" sdf2-mode-syntax-table)
  (modify-syntax-entry ?\' "w" sdf2-mode-syntax-table)
  (modify-syntax-entry ?\\ "\\" sdf2-mode-syntax-table)))
  
(defvar stratego-mode-abbrev-table nil
  "Abbrev table used while in stratego mode.")
(define-abbrev-table 'stratego-mode-abbrev-table ())


(defun stratego-mode ()
  "Major mode for editing Stratego specifications"
  (interactive)
  (kill-all-local-variables)
  (use-local-map stratego-mode-map)
  (set-syntax-table stratego-mode-syntax-table)
  (setq local-abbrev-table stratego-mode-abbrev-table)
  (setq major-mode 'stratego-mode)
  (setq mode-name "Stratego")
  
  (make-local-variable 'indent-line-function)
  (setq indent-line-function 'stratego-indent-line)

  (make-local-variable 'comment-start-skip)
  (setq comment-start-skip "(\\*+ *")
  (make-local-variable 'comment-start)
  (setq comment-start "(* ")
  (make-local-variable 'comment-end)
  (setq comment-end   " *)")
  (make-local-variable 'comment-multi-line)
  (setq comment-multi-line t)
  (make-local-variable 'comment-indent-function)
  (setq comment-indent-function 'stratego-comment-indent)

  (make-local-variable 'font-lock-defaults)
  (setq font-lock-defaults '(stratego-keywords nil nil nil nil))
			     
  (run-hooks 'stratego-mode-hook))

(defun sdf2-mode ()
  "Major mode for editing SDF2 specifications"
  (interactive)
  (kill-all-local-variables)
  (use-local-map stratego-mode-map)
  (set-syntax-table sdf2-mode-syntax-table)
  (setq local-abbrev-table stratego-mode-abbrev-table)
  (setq major-mode 'sdf2-mode)
  (setq mode-name "SDF2")
  
  (make-local-variable 'indent-line-function)
  (setq indent-line-function 'sdf2-indent-line)

  (make-local-variable 'comment-start-skip)
  (setq comment-start-skip "%%+ *")
  (make-local-variable 'comment-start)
  (setq comment-start "%% ")
  (make-local-variable 'comment-end)
  (setq comment-end "")
  (make-local-variable 'comment-multi-line)
  (setq comment-multi-line nil)
  (make-local-variable 'comment-indent-function)
  (setq comment-indent-function 'stratego-comment-indent)

  (make-local-variable 'font-lock-defaults)
  (setq font-lock-defaults '(sdf2-keywords nil nil nil nil))
			     
  (run-hooks 'stratego-mode-hook))

(defvar stratego-keywords
  '("\\<module\\>" "\\<imports\\>" "\\<exports\\>" "\\<rules\\>" "\\<not\\>"
    "\\<operations\\>" "\\<constructors\\>" "\\<stratego\\>" "\\<script\\>" 
    "\\<strategies\\>" "\\<signature\\>" "\\<sorts\\>" "\\<overlays\\>"
    "\\<load\\>" "\\<dump\\>" "\\<one\\>"
    "\\<id\\>" "\\<main\\>"  "\\<rec\\>" "\\<where\\>" "->" "=>" "<=" "!" "?"
    ;("\\<repeat\\>" . font-lock-builtin-face)
    ;("\\<try\\>" . font-lock-builtin-face)
    ;("\\<topdown\\>" . font-lock-builtin-face)
    ;("\\<one\\>" . font-lock-builtin-face)
    ;("\\<oncetd\\>" . font-lock-builtin-face)
    ;("\\<not\\>" . font-lock-builtin-face)
    ("^[ \t]*\\([a-zA-Z0-9'_-]+\\)[ \t]*\\(([^)]*)[ \t]*\\)?:" 
     (1 font-lock-type-face nil nil)
     (2 font-lock-variable-name-face nil t))
    ("^[ \t]*\\([a-zA-Z0-9'_-]+\\)[ \t]*\\(([^)]*)[ \t]*\\)?=" 
     (1 font-lock-function-name-face nil nil)
     (2 font-lock-variable-name-face nil t))
    ))

(defvar sdf2-keywords
  '("\\<module\\>" "\\<imports\\>" "\\<exports\\>" 
    "\\<context-free\\>" "\\<lexical\\>" "\\<syntax\\>" "\\<sorts\\>"
    "\\<restrictions\\>" "\\<definition\\>" "\\<priorities\\>"
    ("\\<cons\\>" . font-lock-function-name-face)
    ("\\<reject\\>" . font-lock-function-name-face)
    ("\\<bracket\\>" . font-lock-function-name-face)
    ("\\<left\\>" . font-lock-function-name-face)
    ("->[ \t]*\\([a-zA-Z0-9'_-]+\\)" .(1 font-lock-variable-name-face nil nil))
  ))

(defun stratego-comment-indent ()
  0)

(defun stratego-indent-line ()
  "Indent current line as Stratego code."
  (interactive)
  (indent-line-to (stratego-get-indent stratego-indentation-list))
  (save-excursion
    (stratego-format-line)))

(defun sdf2-indent-line ()
  "Indent current line as SDF2 code."
  (interactive)
  (indent-line-to (stratego-get-indent sdf2-indentation-list))
  (save-excursion
    (sdf2-format-line)))


(defvar stratego-base-indent 0)
(defvar stratego-empty-indent 0)
(defvar stratego-lvl1-indent 2)
(defvar stratego-rule-indent 2)
(defvar stratego-strategy-indent 2)
(defvar stratego-default-indent 8)
(defvar stratego-colon-indent 6)
(defvar stratego-after-colon-indent 8)
(defvar stratego-equals-indent 6)
(defvar stratego-after-equals-indent 8)
(defvar stratego-arrow-indent 40)
(defvar stratego-after-arrow-indent 42)

(defvar stratego-indentation-list
  '(("module" . stratego-base-indent)
    ("imports" . stratego-base-indent)
    ("rules" . stratego-base-indent)
    ("signature" . stratego-base-indent)
    ("strategies" . stratego-base-indent)
    ("sorts" . stratego-lvl1-indent)
    ("operations" . stratego-lvl1-indent)
    ("[a-zA-Z_()'0-9 \t]*:.*$" . stratego-rule-indent)
    (".*=[^>].*$" . stratego-strategy-indent)
    (".*=$" . stratego-strategy-indent)
    ("[ \t]*$" . stratego-empty-indent)))

(defvar sdf2-indentation-list
  '(("definition" . stratego-base-indent)
    ("module" . stratego-base-indent)
    ("imports" . stratego-base-indent)
    ("exports" . stratego-base-indent)
    ("sorts" . stratego-lvl1-indent)
    ("context-free" . stratego-lvl1-indent)
    ("lexical" . stratego-lvl1-indent)
    ("restrictions" . stratego-lvl1-indent)
    (".*->.*$" . stratego-strategy-indent)
    (".*-/-.*$" . stratego-strategy-indent)
    (">" . stratego-lvl1-indent)
    ("[ \t]*$" . stratego-empty-indent)
    ))

(defun stratego-get-bol ()
  (save-excursion
    (beginning-of-line)
    (point)))
  
(defun stratego-get-eol ()
  (save-excursion
    (end-of-line)
    (point)))
  
(defun stratego-get-indent (list)
  (save-excursion
    (beginning-of-line)
    (skip-chars-forward " \t")
    (let
	((pos nil))
      (while list
	(if (looking-at (car (car list)))
	    (setq pos (cdr (car list))))
	(setq list (cdr list)))
      (if (not pos)
	  (setq pos (or (stratego-last-open-column)
			stratego-default-indent)))
      (eval pos))))

(defun stratego-format-line ()
  (interactive)
  (save-excursion
    (beginning-of-line)
    (if (search-forward ":" (stratego-get-eol) t)
	(progn
	  (backward-char)
	  (indent-to stratego-colon-indent)
	  (forward-char)
	  (if (looking-at "[ \t]*$")
	      (indent-to stratego-after-colon-indent))))
    (if (search-forward-regexp "=[^>]" (stratego-get-eol) t)
	(progn
	  (backward-char 2)
	  (indent-to stratego-equals-indent)
	  (forward-char 2)
	  (if (looking-at "[ \t]*$")
	      (indent-to stratego-after-equals-indent))))
    (if (search-forward-regexp "=$" (stratego-get-eol) t)
	(progn
	  (backward-char)
	  (indent-to stratego-equals-indent)))))

(defun sdf2-format-line ()
  (interactive)
  (save-excursion
    (beginning-of-line)
    (if (search-forward "->" (stratego-get-eol) t)
	(progn
	  (backward-char 2)
	  (indent-to stratego-arrow-indent)
	  (forward-char 2)
	  (if (looking-at "[ \t]*$")
	      (indent-to stratego-after-arrow-indent))))))

(defun stratego-electric-colon (N)
  (interactive "p")
  (save-excursion
    (indent-line-to stratego-rule-indent))
  (indent-to stratego-colon-indent)
  (insert-char ?: N)
  (indent-to stratego-after-colon-indent))

(defun stratego-electric-equals (N)
  (interactive "p")
  (save-excursion
    (indent-line-to stratego-strategy-indent))
  (indent-to stratego-equals-indent)
  (insert-char ?= N)
  (indent-to stratego-after-equals-indent))

(defun stratego-insert-Hs-list ()
  (interactive)
  (insert "Hs([])")
  (backward-char 2))

(defvar stratego-open-paren "[({"
  "Characters which will be considered as open parens.")

(defvar stratego-close-paren "])}"
  "Characters which will be considered as close parens.

`]' must be the first character, if it is used.")

(defun stratego-find-last-open (&optional OPEN CLOSE)
  "Returns the posistion of the last open paren in the buffer."
  (save-excursion
    (if (not OPEN)
	(setq OPEN stratego-open-paren))
    (if (not CLOSE)
	(setq CLOSE stratego-close-paren))
    (let
	((found nil)
	 (pos nil)
	 (nesting 1)
	 char)
      (while (and (not found)
		  (skip-chars-backward (concat "^" OPEN CLOSE))
		  (> (point) 1))
	(backward-char)
	(setq char (buffer-substring-no-properties (point) (+ 1 (point))))
	(if (string-match (concat "[" OPEN "]") char)
	    (progn
	      (setq nesting (- nesting 1))
	      (if (= nesting 0)
		  (setq pos (point) found t))))
	(if (string-match (concat "[" CLOSE "]") char)
	    (setq nesting (+ nesting 1))))
      pos)))

(defun stratego-last-open-column ()
  "Returns the column number of the last open paren, or  nil  if none exists."
  (save-excursion
    (let ((pos (stratego-find-last-open)))
      (if pos
	  (progn
	    (goto-char pos)
	    (+ 1 (current-column)))
	nil))))
	       
	    
    

  
