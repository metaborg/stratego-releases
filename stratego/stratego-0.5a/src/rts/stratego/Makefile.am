includedir = $(prefix)/include/stratego

include_HEADERS = aterm-extension.h  debug.h  options.h  svm.h  util.h

EXTRA_DIST = $(include_HEADERS)


