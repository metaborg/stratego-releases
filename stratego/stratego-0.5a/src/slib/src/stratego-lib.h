/*

Copyright (C) 1998, 1999 Eelco Visser <visser@acm.org>

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2, or (at your option)
any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
02111-1307, USA.

*/

/* $Id: stratego-lib.h,v 1.13 2000/10/30 00:09:53 visser Exp $ */

#ifndef ST_PRIM
#define ST_PRIM

/* Strings */

void *_is_string(void);
void new(void);
void *_implode_string(void);
void *_explode_string(void);

#define _ST_new()            new()
#define _ST_is_string()      GO(_is_string())
#define _ST_implode_string() GO(_implode_string())
#define _ST_explode_string() GO(_explode_string())

/* Numbers */

void *_is_int(void);
void *_is_real(void);
void *_int(void);
void *_plus(void);
void *_minus(void);
void *_mul(void);
void *_div(void);
void *_mod(void);
void *_geq(void);
void *_leq(void);
void *_gt(void);
void *_lt(void);
void *_max(void);
void *_min(void);

void *_sqrt(void);
void *_sin(void);
void *_cos(void);

#define _ST_int()            GO(_int())
#define _ST_is_int()         GO(_is_int())
#define _ST_is_real()        GO(_is_real())

#define _ST_plus()           GO(_plus())
#define _ST_add()            GO(_plus())
#define _ST_minus()          GO(_minus())
#define _ST_subt()           GO(_minus())
#define _ST_mul()            GO(_mul())
#define _ST_div()            GO(_div())
#define _ST_mod()            GO(_mod())
#define _ST_geq()            GO(_geq())
#define _ST_leq()            GO(_leq())
#define _ST_gt()             GO(_gt())
#define _ST_lt()             GO(_lt())
#define _ST_max()            GO(_max())
#define _ST_min()            GO(_min())

#define _ST_sqrt()           GO(_sqrt())
#define _ST_sin()            GO(_sin())
#define _ST_cos()            GO(_cos())

/* Terms */

void *_mkterm(void);
void *_explode_term(void);
void *_address_lt(void);
void *_address(void);

#define _ST_mkterm()         GO(_mkterm())
#define _ST_explode_term()   GO(_explode_term())
#define _ST_address_lt()     GO(_address_lt())
#define _ST_address()        GO(_address())

/* Kids: map constructor application to tuple of children */

ATerm mktuple(ATermList l);
ATbool is_tuple(ATerm t);
ATbool is_empty_tuple(ATerm t);
ATbool is_empty_tuple(ATerm t);
void kids(void);

/* IO */

void *ST_file_exists(void);
void *ST_open_file(void);
void *ST_print(void);
void *ST_printascii(void);
void *ST_printnl(void);
void *ST_ReadFromFile(void);
void *ST_WriteToBinaryFile(void);
void *ST_WriteToTextFile(void);
void *ST_append_file(void);
void *ST_close_file(void);


void *ST_PrintStack(void);

#define _ST_file_exists()       GO(ST_file_exists())
#define _ST_open_file()         GO(ST_open_file())
#define _ST_print()             GO(ST_print())
#define _ST_printnl()           GO(ST_printnl())
#define _ST_printascii()        GO(ST_printascii())
#define _ST_ReadFromFile()      GO(ST_ReadFromFile())
#define _ST_WriteToBinaryFile() GO(ST_WriteToBinaryFile())
#define _ST_WriteToTextFile()   GO(ST_WriteToTextFile())
#define _ST_append_file()       GO(ST_append_file())
#define _ST_close_file()        GO(ST_close_file())


#define _ST_PrintStack()        GO(ST_PrintStack())

/* exec */

void *ST_exit(void);
void *ST_call(void);
void *ST_call_noisy(void);
void *ST_get_pid(void);

#define _ST_exit()       GO(ST_exit())
#define _ST_call()       GO(ST_call())
#define _ST_call_noisy() GO(ST_call_noisy())
#define _ST_get_pid()    GO(ST_get_pid())

/* time */

void *ST_dtime(void);
#define _ST_dtime() GO(ST_dtime())

#include "stratego/tables.h"

#endif
