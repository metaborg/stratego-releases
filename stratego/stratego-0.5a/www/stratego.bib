% BibTeX bibliography file
		  
@TechReport{JV99,
  author =	 {Johann, Patricia and Visser, Eelco},
  title =	 {Warm Fusion in {Stratego}: {A} Case Study in the
                  Generation of Program Transformation Systems},
  institution =	 {Department of Computer Science, Universiteit
                  Utrecht},
  address =	 {Utrecht, The Netherlands},
  year =	 1999
}

@InProceedings{LV97,
  author =	 {Luttik, Bas and Visser, Eelco},
  title =	 {Specification of Rewriting Strategies},
  booktitle =	 {2nd International Workshop on the Theory and
                  Practice of Algebraic Specifications (ASF+SDF'97)},
  editor =	 {M. P. A. Sellink},
  year =	 {1997},
  month =	 {November},
  series =	 {Electronic Workshops in Computing},
  publisher =	 {Springer-Verlag},
  address =	 {Berlin},
  URL =		 {http://www.springer.co.uk/ewic/workshops/ASFSDF97/}
}

@Article{VB98,
  author =	 {Visser, Eelco and Benaissa, Zine-el-Abidine},
  title =	 {A Core Language for Rewriting},
  journal =	 {Electronic Notes in Theoretical Computer Science},
  year =	 1998,
  volume =	 15,
  month =	 {September},
  note =	 {In C.~Kirchner and H.~Kirchner, editors, Proceedings
                  of the Second International Workshop on Rewriting
                  Logic and its Applications (WRLA'98),
                  Pont-{\`a}-Mousson, France}
}

@Article{VBT98,
  author =	 {Visser, Eelco and Benaissa, Zine-el-Abidine and
                  Tolmach, Andrew},
  title =	 {Building Program Optimizers with Rewriting
                  Strategies},
  journal =	 {ACM SIGPLAN Notices},
  year =	 1999,
  volume =	 34,
  number =	 1,
  pages =	 {13--26},
  month =	 {January},
  note =	 {Proceedings of the International Conference on
                  Functional Programming (ICFP'98)}
}

@TechReport{Vis2000,
  author =	 {Eelco Visser},
  title =	 {Language Independent Traversals for Program
                  Transformation},
  institution =	 {Department of Computer Science, Universiteit
                  Utrecht},
  year =	 2000,
  address =	 {Utrecht, The Netherlands}
}

@InProceedings{Vis99,
  author =	 {Visser, Eelco},
  title =	 {Strategic Pattern Matching},
  booktitle =	 {Rewriting Techniques and Applications (RTA'99)},
  pages =	 {30--44},
  year =	 1999,
  editor =	 {P. Narendran and M. Rusinowitch},
  volume =	 1631,
  series =	 {Lecture Notes in Computer Science},
  address =	 {Trento, Italy},
  month =	 {July},
  publisher =	 {Springer-Verlag}
}

@TechReport{Vis99.compiler,
  author =	 "Visser, Eelco",
  title =	 "The {Stratego} Compiler",
  institution =	 "Department of Computer Science, Universiteit
                  Utrecht",
  year =	 1999,
  address =	 "Utrecht, The Netherlands"
}

@Manual{Vis99.lib,
  title =	 "The {Stratego} Library",
  author =	 "Visser, Eelco",
  organization = "Department of Computer Science, Universiteit
                  Utrecht",
  address =	 "Utrecht, The Netherlands",
  year =	 1999
}

@Manual{Vis99.refman,
  title =	 "The {Stratego} Reference Manual",
  author =	 "Visser, Eelco",
  organization = "Department of Computer Science, Universiteit
                  Utrecht",
  address =	 "Utrecht, The Netherlands",
  year =	 1999
}

@Manual{Vis99.tutorial,
  title =	 "The {Stratego} Tutorial",
  author =	 "Visser, Eelco",
  organization = "Department of Computer Science, Universiteit
                  Utrecht",
  address =	 "Utrecht, The Netherlands",
  year =	 1999
}

