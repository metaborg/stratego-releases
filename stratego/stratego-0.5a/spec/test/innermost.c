#include <stratego.h>
#include <stratego-lib.h>
#include <stratego-ext.h>

VARDECL(AFun,z_0);
VARDECL(AFun,y_0);
VARDECL(AFun,w_0);
PROCDECL(_stratego_main);

PROC(_stratego_init_afuns)
  MOVE(z_0,ATmakeAFun("Succ",1,0));
  MOVE(y_0,ATmakeAFun("Zero",0,0));
  MOVE(w_0,ATmakeAFun("Plus",2,0));
ENDPROC

PROC(_stratego_main)
  Rpush(a_0);
  p_0 :
  AllInit();
  b_0 :
  AllNextSon(&&c_0);
  Rpush(e_0);
  goto p_0;
  e_0 :
  goto b_0;
  c_0 :
  AllBuild();
  Rpush(k_0);
  q_0 :
  Cpush(m_0);
  Epushd(0,4);
  MoveTop(0,1);
  goto d_0;
  d_0 :
  TestFunFC(w_0,&&fail,Egetd(0,1));
  MoveArg(0,2,0,1,0);
  MoveArg(0,4,0,1,1);
  goto f_0;
  f_0 :
  TestFunFC(y_0,&&x_0,Egetd(0,2));
  Rpush(v_0);
  goto r_0;
  x_0 :
  TestFunFC(z_0,&&fail,Egetd(0,2));
  MoveArg(0,3,0,2,0);
  Rpush(v_0);
  goto s_0;
  v_0 :
  goto u_0;
  s_0 :
  Epushd(1,2);
  NotNULLd(0,3);
  NotNULLd(0,4);
  Tset(App2("Plus",Egetd(0,3),Egetd(0,4)));
  Rpush(a_1);
  goto q_0;
  a_1 :
  MoveTop(1,2);
  Move(1,1,1,2);
  NotNULLd(1,1);
  Tset(App1("Succ",Egetd(1,1)));
  Rpush(b_1);
  goto q_0;
  b_1 :
  Epopd(1,2);
  Return();
  u_0 :
  goto t_0;
  r_0 :
  NotNULLd(0,4);
  Tset(Egetd(0,4));
  Return();
  t_0 :
  Epopd(0,4);
  goto l_0;
  m_0 :
  goto c_1;
  l_0 :
  Cpop();
  c_1 :
  Return();
  k_0 :
  Return();
  a_0 :
ENDPROC

DOIT

