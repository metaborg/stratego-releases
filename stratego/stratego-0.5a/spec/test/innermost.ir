SDef("main",[],
Rec("im",
 Seq(
  All(Call(SVar("im"),[])),

  Rec("red",
   LChoice(
    Seq(
     Test(
      All(
       Rec("q_1",
        Seq(
         All(Call(SVar("q_1"),[])),
         Not(
          Choice(
           Scope(["j_1"],
            Seq(Match(Op("Plus",[Op("Zero",[]),Var("j_1")])),
                Build(Var("j_1")))),
           Scope(["k_1","l_1"],
            Seq(Match(Op("Plus",[Op("Succ",[Var("k_1")]),Var("l_1")])),
                Build(Op("Succ",[Op("Plus",[Var("k_1"),Var("l_1")])])))))))))),

     Seq(

      Choice(

       Scope(["m_1"],
        Seq(Match(Op("Plus",[Op("Zero",[]),Var("m_1")])),
            Build(Var("m_1")))),

       Scope(["n_1","o_1"],
        Seq(Match(Op("Plus",[Op("Succ",[Var("n_1")]),Var("o_1")])),
            Build(Op("Succ",[Op("Plus",[Var("n_1"),Var("o_1")])]))))),

      Rec("r_1",
       Seq(
        All(Call(SVar("r_1"),[])),
        Call(SVar("red"),[])))))

     ,Id))
  ))
)
