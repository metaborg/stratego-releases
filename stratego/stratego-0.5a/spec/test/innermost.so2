[SDef("main",[],Rec("t_0",Seq(All(Call(SVar("t_0"),[])),Rec("x_0",LChoice(Choice(Scope(["u_0"],Seq(Match(Op("Plus",[Op("Zero",[]),Var("u_0")])),Build(Var("u_0")))),Scope(["v_0","w_0"],Seq(Match(Op("Plus",[Op("Succ",[Var("v_0")]),Var("w_0")])),Seq(Scope(["e_0"],Seq(Seq(Build(Op("Plus",[Var("v_0"),Var("w_0")])),Seq(Call(SVar("x_0"),[]),Seq(Match(Var("e_0")),Id))),Build(Op("Succ",[Var("e_0")])))),Call(SVar("x_0"),[]))))),Id)))))]
