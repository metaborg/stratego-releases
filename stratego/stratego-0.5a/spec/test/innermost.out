SDef("main",[],

Rec("im",Seq(All(Call(SVar("im"),[])),Rec("red",LChoice(Seq(Test(All(Rec("q_1",Seq(All(Call(SVar("q_1"),[])),Not(Choice(Scope(["j_1"],Seq(Match(Op("Plus",[Op("Zero",[]),Var("j_1")])),Build(Var("j_1")))),Scope(["k_1","l_1"],Seq(Match(Op("Plus",[Op("Succ",[Var("k_1")]),Var("l_1")])),Build(Op("Succ",[Op("Plus",[Var("k_1"),Var("l_1")])])))))))))),Choice(Scope(["m_1"],Seq(Match(Op("Plus",[Op("Zero",[]),Var("m_1")])),BA(Rec("r_1",Seq(All(Call(SVar("r_1"),[])),Call(SVar("red"),[]))),Var("m_1")))),Scope(["n_1","o_1"],Seq(Match(Op("Plus",[Op("Succ",[Var("n_1")]),Var("o_1")])),BA(Call(SVar("red"),[]),Op("Succ",[BA(Call(SVar("red"),[]),Op("Plus",[BA(Rec("r_1",Seq(All(Call(SVar("r_1"),[])),Call(SVar("red"),[]))),Var("n_1")),BA(Rec("r_1",Seq(All(Call(SVar("r_1"),[])),Call(SVar("red"),[]))),Var("o_1"))]))])))))),Id)))))
